### ISSUE TITLE
Aliases with “and-or list” value is highlighted as error

### ISSUE THREAD

**Issue by FranklinYu** [zsh-users/zsh-syntax-highlighting#769]
After

```shell
alias ls1=ls ls2='ls1 && ls1' ls3='ls1 || ls1'
```

then `ls2` and `ls3` will be highlighted as error (red). When I actually run `ls2` or `ls3`, Zsh correctly find the `ls` and run it twice (listing the `$PWD` twice), indicating that the syntax should be correct.

### environment

operating system: macOS 10.15.6
Zsh: 5.7.1 (macOS system version)
z-sy-h: 0.7.1 (installed from MacPorts)

**Comment by phy1729** [comment]
The and-or list (in zsh terms a sublist) seems to be a red herring. Both `alias a=ls b='a; a'` and `alias a=ls b='a | a'` result in `b` being highlighted red. This is due to loop detection not being reset on the start of a new simple command.

```
diff --git a/highlighters/main/main-highlighter.zsh b/highlighters/main/main-highlighter.zsh
index e59c61c..2a29d0d 100644
--- a/highlighters/main/main-highlighter.zsh
+++ b/highlighters/main/main-highlighter.zsh
@@ -879,6 +879,7 @@ _zsh_highlight_main_highlighter_highlight_list()
         next_word=':start:'
         highlight_glob=true
         saw_assignment=false
+        seen_alias=()
         if [[ $arg != '|' && $arg != '|&' ]]; then
           next_word+=':start_of_pipeline:'
         fi
```
appears to fix the issue. Will confirm and add tests tomorrow.

**Comment by FranklinYu** [comment]
I can verify that the issue is fixed in master branch. Look forward to next release!

**Comment by danielshahaf** [comment]
At this point in time master is stable, so feel free to upgrade to it.  (That's not the same thing as "always use latest master"; I'm just suggesting that you upgrade to a snapshot of current master, if you need this fix.)

**Comment by CVirus** [comment]
This fix introduced an issue for me. I have `alias tmux="cd /somedir && tmux"` in my `.zshrc` now when I try to run to type `tmux` on the command line, it stops at the letter `u` and wouldn't write the `x` for some reason. The letter x is shown as gray as if it is going to be autocomplete and I can't type or backspace anything. I have to CTRL + c
![DeepinScreenshot_select-area_20201124232648](https://user-images.githubusercontent.com/106264/100158533-9ee51d00-2eac-11eb-9911-1c33076f3aae.png)

**Comment by phy1729** [comment]
Probably a duplicate of #775. Does #776 fix the issue? It hasn't been merged because it breaks 4.3.11, but I haven't had time to figure out why yet.

**Comment by danielshahaf** [comment]
> Probably a duplicate of #775 
> <https://github.com/zsh-users/zsh-syntax-highlighting/issues/775>. Does 
> #776 <https://github.com/zsh-users/zsh-syntax-highlighting/pull/776> 
> fix the issue? It hasn't been merged because it breaks 4.3.11, but I 
> haven't had time to figure out why yet.

Might be the precommand array assignment («in_array=() foo»)?

Also, note that «! ${+array[1]}» is faster than «$#array == 0» (O(1) v. O(N)).
